primitives/src/market.rs

```diff
@@ -56,9 +56,58 @@ pub struct Market<AI, BN, M> {
     pub resolved_outcome: Option<OutcomeReport>,
     /// See [`MarketDisputeMechanism`].
     pub dispute_mechanism: MarketDisputeMechanism,
+    /// The bonds reserved for this market.
+    pub bonds: MarketBonds<AI, Balance>,
 }
 
-impl<AI, BN, M> Market<AI, BN, M> {
+/// Tracks the status of a bond.
+#[derive(Clone, Decode, Encode, MaxEncodedLen, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct Bond<AccountId, Balance> {
+    /// The account that reserved the bond.
+    pub who: AccountId,
+    /// The amount reserved.
+    pub value: Balance,
+    /// `true` if and only if the bond is unreserved and/or (partially) slashed.
+    pub is_settled: bool,
+}
+
+impl<AccountId, Balance> Bond<AccountId, Balance> {
+    pub fn new(who: AccountId, value: Balance) -> Bond<AccountId, Balance> {
+        Bond { who, value, is_settled: false }
+    }
+}
+
+/// Tracks bonds associated with a prediction market.
+#[derive(Clone, Decode, Encode, MaxEncodedLen, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct MarketBonds<AccountId, Balance> {
+    pub advisory: Option<Bond<AccountId, Balance>>,
+    pub oracle: Option<Bond<AccountId, Balance>>,
+    pub validity: Option<Bond<AccountId, Balance>>,
+}
```

Comment:
The same question occurs when thinking about where to store the bond for outside oracles. Use the `oracle` field (we could slash the oracle when an outsider reports, although that's not the case right now) or add another field?

Is there any advantage to using fewer fields than storage conservation? If not, then I'm in favor of keeping the separate bonds. What we gain by using extra fields is that bond settlements are more clearly abstracted. If we use the field for both advisory and validity bond, we tie the meaning of that field to the rest of the state of the `Market` object. In terms of future-proofing and migration-safety, that's not ideal. For example:

- If we change the `AdvisoryBond`, we need to check the status of the market before changing the `market_bond` in the migration.
- If we want to slash all bonds, as we do in `admin_destroy_market`, then we need to check the status: If the market is proposed, slash that bond, other wise this one.

Reply:
Good point Harald. It could be done by a `bonds: BTreeMap`, but I am okay with it right now too. It's just clear to name all possible bonds. One unnecessary field with `None` does not take so much space, I guess. You could also use an `enum` which has struct fields for the possible bond combinations. That could be even clearer.

Reply:
But that would tie market state and bonds even closer together. This is how settling a bond would look like then, wouldn't it?

```rust
match market.market_creation {
    MarketCreation::Permissionless => {
        match market.bonds {
            MarketCreationBonds::Advised => {
                // error
            },
            MarketCreationBonds::Permissionless => {
                // ok
            }
        }
    },
    // Symmetric for `Advised`
}
```

That's what I'm trying to move away from.

Reply:
I see. Much indentation and double naming here. Yeah, as I said, I am okay with the current approach.

Reply:
The gain would be more storage efficiency (this is neglectable) and easier handling of the bond. The loss would be transparency: To know what kind of bond this is we have to query the market creation type. Some examples how the code would reduce by uniting them:

<table>
<tr>
<th>Before</th>
<th>After</th>
<th>After (compact)</th>
</tr>
<tr>
<td>
  
```rust
match &market.bonds.advisory {
    Some(bond) if !bond.is_settled => {
        Self::slash_advisory_bond(&market_id, true)?;
    }
    _ => (),
};
match &market.bonds.oracle {
    Some(bond) if !bond.is_settled => {
        Self::slash_oracle_bond(&market_id)?;
    }
    _ => (),
};
match &market.bonds.market_creation {
    Some(bond) if !bond.is_settled => {
        Self::slash_validity_bond(&market_id)?;
    }
    _ => (),
};
```
  
</td>
<td>

```rust
match &market.bonds.oracle {
    Some(bond) if !bond.is_settled => {
        Self::slash_oracle_bond(&market_id)?;
    }
    _ => (),
};
match &market.bonds.market_creation {
    Some(bond) if !bond.is_settled => {
        Self::slash_market_creation_bond(&market_id)?;
    }
    _ => (),
};
```

</td>
<td>

```rust
if let Some(bond) = market.bonds.oracle && !bond.is_settled {
    Self::slash_oracle_bond(&market_id)?;
}
if let Some(bond) = market.bonds.market_creation && !bond.is_settled {
    Self::slash_market_creation_bond(&market_id)?;
}
```

</td>
</tr>
</table>

<table>
<tr>
<th>Before</th>
<th>After</th>
</tr>
<tr>
<td>
  
```rust
let bonds = match creation {
    MarketCreation::Permissionless => MarketBonds {
        advisory: None,
        oracle: Some(Bond::new(sender.clone(), T::OracleBond::get())),
        validity: Some(Bond::new(sender.clone(), T::ValidityBond::get())),
    },
    MarketCreation::Advised => MarketBonds {
        advisory: Some(Bond::new(sender.clone(), T::AdvisoryBond::get())),
        oracle: Some(Bond::new(sender.clone(), T::OracleBond::get())),
        validity: None,
    },
};
```
  
</td>
<td>

```rust
let bond_amount = match old_market.creation {
    MarketCreation::Advised => T::AdvisoryBond::get(),
    MarketCreation::Permissionless => T::ValidityBond::get(),
};
let bonds = MarketCreation::Advised => MarketBonds {
    market_creation: Some(Bond::new(sender.clone(), bond_amount)),
    oracle: Some(Bond::new(sender.clone(), T::OracleBond::get())),
},
```

</td>
</tr>
</table>

The validity and advisory fields are already tied to the market in the current situation. Somewhere in the prediction markets pallet we have to check market conditions to know which of those field to set. However, once set, we can rely on that in the future: The market information was bound into the structure (i.e. when `advisory` is set, we know that the underlying market is an advised market and don't have to check that again in the future). This can reduce the amount of checks required in the future.

I am totally fine to go with the current approach, it just seemed weird to me that at all times only 2 of the 3 fields in the struct are used, that seemed to me like a design error. This is heavy nitpicking here :)

Reply:
> I am totally fine to go with the current approach, it just seemed weird to me that at all times only 2 of the 3 fields in the struct are used, that seemed to me like a design error. This is heavy nitpicking here :)

I take these things seriously, and don't believe this is a nitpick.

> The validity and advisory fields are already tied to the market in the current situation. Somewhere in the prediction markets pallet we have to check market conditions to know which of those field to set.

This seems to be the key observation. Basicallly, what the use of `advisory` and `validity` does is that it double-books the `MarketCreation`. Which has the advantage that a function like `slash_advisory_bond` can be tailored specifically to the needs of that particular bond (`slash_all` parameter).

Nevertheless, I've tried your suggestion and admit that it's just plain better. Basically everything is tighter this way, _except_ that we now have a second optional parameter in the `slash_*_bond` function (see last paragraph).

You might ask why the fields of `MarketBonds` are still optional. Strictly speaking it would be better to have them non-optional, but this would mean more duplicated code once other bonds are introduced which are non-optional (the bond for outsider reports). However, I don't see any design problems originating from this. Yes, someone could somewhere set a non-optional bond to `None`. But that someome could also just change the value of the bond, regardless of whether or not it's optional.

Unfortunately, the compact version of the `if let` expression doesn't work because that's still an unstable feature.

Reply:
[b6807d9](https://github.com/zeitgeistpm/zeitgeist/pull/892/commits/b6807d9bfcaeae07bb4a1ddbf05f39eafc5361ec)